my/core/query_range.py

```diff
@@ -0,0 +1,526 @@
+    assert res == [_Float(2.5), _Float(3.5), _Float(4.5), _Float(5.5)]
+
+
+def test_range_predicate() -> None:
+
+    from functools import partial
+
+    def src() -> Iterator[str]:
+        yield from map(str, range(15))
+
+    identity = lambda o: o
+
+    # convert any float values to ints
+    coerce_int_parser = lambda o: int(float(o))
+    int_filter_func = partial(_create_range_filter, attr_func=identity, end_parser=coerce_int_parser,
+            within_parser=coerce_int_parser, value_coercion_func=coerce_int_parser)
+
+    # filter from 0 to 5
+    rn: Optional[RangeTuple] = RangeTuple("0", "5", None)
+    zero_to_five_filter: Optional[Where] = int_filter_func(unparsed_range=rn)
+    assert zero_to_five_filter is not None
+    # this is just a Where function, given some input it return True/False if the value is allowed
+    assert zero_to_five_filter(3) is True
+    assert zero_to_five_filter(10) is False
+
+    # this is expected, range_predicate is not inclusive on the far end
+    assert list(filter(zero_to_five_filter, src())) == ["0", "1", "2", "3", "4"]
+
+    # items less than 3, going 3.5 (converted to 3 by the coerce_int_parser) down
+    rn = RangeTuple(None, 3, "3.5")
+    assert list(filter(int_filter_func(unparsed_range=rn, attr_func=identity), src())) == ["0", "1", "2"]
+
+def test_parse_range() -> None:
+
+    from functools import partial
+
+    import pytest
+
+    rn = RangeTuple("0", "5", None)
+    res = _parse_range(unparsed_range=rn, end_parser=int, within_parser=int)
+
+    assert res == RangeTuple(after=0, before=5, within=None)
+
+    dt_parse_range = partial(_parse_range, end_parser=parse_datetime_float, within_parser=parse_timedelta_float)
+
+    start_date = datetime.now()
+    end_date = start_date + timedelta(seconds=60)
+
+    # convert start items to strings, which need to be parsed back
+    rn = RangeTuple(str(start_date), str(end_date.timestamp()), None)
+    res2 = dt_parse_range(unparsed_range=rn)
+
+    assert res2 == RangeTuple(after=start_date.timestamp(), before=end_date.timestamp(), within=None)
+
+    # cant specify all three
+    with pytest.raises(QueryException, match=r"Cannot specify 'after', 'before' and 'within'"):
+        dt_parse_range(unparsed_range=RangeTuple(str(start_date), str(end_date.timestamp()), "7d"))
+
+    # if you specify noting, should return None
+    res3 = dt_parse_range(unparsed_range=RangeTuple(None, None, None))
+    assert res3 is None
+
+
+def test_parse_timedelta_string() -> None:
+
+    import pytest
+
+    with pytest.raises(ValueError, match=r"Could not parse time duration from"):
+        parse_timedelta_string("5xxx")
+
+    res = parse_timedelta_string("1w5d5h10m50s")
+    assert res == timedelta(days=7.0 + 5.0, hours=5.0, minutes=10.0, seconds=50.0)
+
+
+def test_parse_datetime_float() -> None:
+
+    pnow = parse_datetime_float("now")
+    sec_diff = abs((pnow - datetime.now().timestamp()))
+    # should probably never fail? could mock time.time
```

Comment:
eh.. IMO good enough, mocking time seems a bit insane  xD

Reply:
yeah, didnt want to add the extra dependency either, there are some third part libs that can handle this